sentinel-extension/sentinel-datasource-eureka/src/main/java/com/alibaba/csp/sentinel/datasource/eureka/EurekaDataSource.java

```diff
@@ -0,0 +1,215 @@
+
+import com.alibaba.csp.sentinel.datasource.AutoRefreshDataSource;
+import com.alibaba.csp.sentinel.datasource.Converter;
+import com.alibaba.csp.sentinel.datasource.ReadableDataSource;
+import com.alibaba.csp.sentinel.log.RecordLog;
+import com.alibaba.csp.sentinel.util.AssertUtil;
+import com.alibaba.csp.sentinel.util.StringUtil;
+import com.alibaba.fastjson.JSON;
+
+import java.io.*;
+import java.net.HttpURLConnection;
+import java.net.InetAddress;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * <p>
+ * A {@link ReadableDataSource} based on Eureka. This class will automatically
+ * fetches the metadata of the instance every period.
+ * </p>
+ * <p>
+ * Limitations: Default refresh interval is 10s. Because there is synchronization between eureka servers,
+ * it may take longer to take effect.
+ * </p>
+ *
+ * @author: liyang
+ * @create: 2020-05-23 12:01
+ */
+public class EurekaDataSource<T> extends AutoRefreshDataSource<String, T> {
+
+    private static final long DEFAULT_REFRESH_MS = 10000;
+
+    /**
+     * connect timeout: 3s
+     */
+    private static final int DEFAULT_CONNECT_TIMEOUT_MS = 3000;
+
+    /**
+     * read timeout: 30s
+     */
+    private static final int DEFAULT_READ_TIMEOUT_MS = 30000;
+
+
+    private int connectTimeoutMills;
+
+
+    private int readTimeoutMills;
+
+    /**
+     * eureka instance appid
+     */
+    private String appId;
+    /**
+     * eureka instance id
+     */
+    private String instanceId;
+
+    /**
+     * collect of eureka server urls
+     */
+    private List<String> serviceUrls;
+
+    /**
+     * metadata key of the rule source
+     */
+    private String ruleKey;
+
+
+    public EurekaDataSource(String appId, String instanceId, List<String> serviceUrls, String ruleKey,
+                            Converter<String, T> configParser) {
+        this(appId, instanceId, serviceUrls, ruleKey, configParser, DEFAULT_REFRESH_MS, DEFAULT_CONNECT_TIMEOUT_MS, DEFAULT_READ_TIMEOUT_MS);
+    }
+
+
+    public EurekaDataSource(String appId, String instanceId, List<String> serviceUrls, String ruleKey,
+                            Converter<String, T> configParser, long refreshMs, int connectTimeoutMills,
+                            int readTimeoutMills) {
+        super(configParser, refreshMs);
+        AssertUtil.notNull(appId, "appId can't be null");
+        AssertUtil.notNull(instanceId, "instanceId can't be null");
+        AssertUtil.notNull(ruleKey, "ruleKey can't be null");
+        AssertUtil.notNull(configParser, "configParser can't be null");
+        AssertUtil.assertState(refreshMs > 0, "refreshMs must be greater than 0");
+        AssertUtil.assertState(connectTimeoutMills > 0, "connectTimeoutMills must be greater than 0");
+        AssertUtil.assertState(readTimeoutMills > 0, "readTimeoutMills must be greater than 0");
+        AssertUtil.assertNotEmpty(serviceUrls, "serviceUrls can't be empty");
```

Comment:
#Line100,#Line101 can be removed since the super constructor has checked their validation.
Valid the parameters in order seems better reading : )

Reply:
Good idea!